### PR TITLE
[posix] make base class destructor virtual

### PR DESCRIPTION
**Pull request by rxwen**
- get rid of compiler warning

### CODE REVIEW

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1: 

Agree with that.

File: src/posix/platform/mainloop.hpp [hunk 1]
```diff
@@ -65,6 +65,12 @@ class Source
      */
     virtual void Process(const otSysMainloopContext &aContext) = 0;
 
+    /**
+     * This method mark desturctor virtual.
+     *
+     */
+    virtual ~Source() = 0;
```
Comment (superwhd) [inline src/posix/platform/mainloop.hpp:72]: I think this should be `= default` rather than `= 0`?

**Comment by jwhui** [comment]
@rxwen , the docker builds seem to be failing:
```
#9 65.05   CXX      ot_ctl-client.o
#9 65.05   CC       ot_daemon-main.o
#9 65.15   CCLD     ot-daemon
#9 65.20   CXXLD    ot-ctl
#9 66.43 ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-udp.o): In function `ot::Posix::Udp::Get()':
#9 66.44 /root/src/openthread/src/posix/../../src/posix/platform/udp.cpp:611: undefined reference to `__cxa_guard_acquire'
#9 66.44 /root/src/openthread/src/posix/../../src/posix/platform/udp.cpp:611: undefined reference to `__cxa_guard_release'
#9 66.44 ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-udp.o): In function `ot::Posix::Udp::~Udp()':
#9 66.44 /root/src/openthread/src/posix/../../src/posix/platform/udp.hpp:37: undefined reference to `operator delete(void*)'
#9 66.44 ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-daemon.o): In function `ot::Posix::Daemon::Get()':
#9 66.44 /root/src/openthread/src/posix/../../src/posix/platform/daemon.cpp:364: undefined reference to `__cxa_guard_acquire'
#9 66.44 /root/src/openthread/src/posix/../../src/posix/platform/daemon.cpp:364: undefined reference to `__cxa_guard_release'
#9 66.44 ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-daemon.o): In function `ot::Posix::Daemon::~Daemon()':
#9 66.44 /root/src/openthread/src/posix/../../src/posix/platform/daemon.hpp:39: undefined reference to `operator delete(void*)'
#9 66.44 collect2: error: ld returned 1 exit status
#9 66.44 Makefile:666: recipe for target 'ot-daemon' failed
#9 66.44 make[5]: *** [ot-daemon] Error 1
#9 66.44 Makefile:812: recipe for target 'all-recursive' failed
#9 66.44 make[4]: *** [all-recursive] Error 1
#9 66.44 Makefile:434: recipe for target 'all-recursive' failed
#9 66.44 make[3]: *** [all-recursive] Error 1
#9 66.45 Makefile:587: recipe for target 'all-recursive' failed
#9 66.45 make[2]: *** [all-recursive] Error 1
#9 66.45 Makefile:513: recipe for target 'all' failed
#9 66.45 make[1]: *** [all] Error 2
#9 66.45 src/posix/Makefile-posix:326: recipe for target 'do-build-posix' failed
#9 66.45 make: *** [do-build-posix] Error 2
#9 ERROR: process "/bin/sh -c cd ~/src/openthread && make -f src/posix/Makefile-posix DAEMON=1" did not complete successfully: exit code: 2
```
https://github.com/openthread/openthread/runs/4434509000?check_suite_focus=true

**Comment by rxwen** [comment]
> @rxwen , the docker builds seem to be failing:
> 
> ```
> #9 65.05   CXX      ot_ctl-client.o
> #9 65.05   CC       ot_daemon-main.o
> #9 65.15   CCLD     ot-daemon
> #9 65.20   CXXLD    ot-ctl
> #9 66.43 ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-udp.o): In function `ot::Posix::Udp::Get()':
> #9 66.44 /root/src/openthread/src/posix/../../src/posix/platform/udp.cpp:611: undefined reference to `__cxa_guard_acquire'
> #9 66.44 /root/src/openthread/src/posix/../../src/posix/platform/udp.cpp:611: undefined reference to `__cxa_guard_release'
> #9 66.44 ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-udp.o): In function `ot::Posix::Udp::~Udp()':
> #9 66.44 /root/src/openthread/src/posix/../../src/posix/platform/udp.hpp:37: undefined reference to `operator delete(void*)'
> #9 66.44 ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-daemon.o): In function `ot::Posix::Daemon::Get()':
> #9 66.44 /root/src/openthread/src/posix/../../src/posix/platform/daemon.cpp:364: undefined reference to `__cxa_guard_acquire'
> #9 66.44 /root/src/openthread/src/posix/../../src/posix/platform/daemon.cpp:364: undefined reference to `__cxa_guard_release'
> #9 66.44 ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-daemon.o): In function `ot::Posix::Daemon::~Daemon()':
> #9 66.44 /root/src/openthread/src/posix/../../src/posix/platform/daemon.hpp:39: undefined reference to `operator delete(void*)'
> #9 66.44 collect2: error: ld returned 1 exit status
> #9 66.44 Makefile:666: recipe for target 'ot-daemon' failed
> #9 66.44 make[5]: *** [ot-daemon] Error 1
> #9 66.44 Makefile:812: recipe for target 'all-recursive' failed
> #9 66.44 make[4]: *** [all-recursive] Error 1
> #9 66.44 Makefile:434: recipe for target 'all-recursive' failed
> #9 66.44 make[3]: *** [all-recursive] Error 1
> #9 66.45 Makefile:587: recipe for target 'all-recursive' failed
> #9 66.45 make[2]: *** [all-recursive] Error 1
> #9 66.45 Makefile:513: recipe for target 'all' failed
> #9 66.45 make[1]: *** [all] Error 2
> #9 66.45 src/posix/Makefile-posix:326: recipe for target 'do-build-posix' failed
> #9 66.45 make: *** [do-build-posix] Error 2
> #9 ERROR: process "/bin/sh -c cd ~/src/openthread && make -f src/posix/Makefile-posix DAEMON=1" did not complete successfully: exit code: 2
> ```
> 
> https://github.com/openthread/openthread/runs/4434509000?check_suite_focus=true

@jwhui Thanks. i noticed it and was trying to figure out the root cause. 

Looks like I have to link against linstdc++ after I added the default destructor.

File: src/posix/platform/mainloop.hpp [hunk 2]
```diff
@@ -65,6 +65,12 @@ class Source
      */
     virtual void Process(const otSysMainloopContext &aContext) = 0;
 
+    /**
+     * This method mark desturctor virtual.
```
Comment (LuDuda) [inline src/posix/platform/mainloop.hpp:69]:
```suggestion
     * This method marks destructor virtual.
```

File: src/posix/Makefile-posix [hunk 3]
```diff
@@ -94,7 +94,10 @@ configure_OPTIONS                   = \
 # Platform specific switches
 
 ifeq ($(DAEMON),1)
-configure_OPTIONS              += --enable-posix-daemon
+configure_OPTIONS              += \
+	--enable-posix-daemon 		  \
+	LIBS=-lstdc++				  \
```
Comment (LuDuda) [inline src/posix/Makefile-posix:99]: nit: alignment

Comment (rxwen) [reply src/posix/Makefile-posix:99]: > nit: alignment

Excuse me, I failed to find the difference in you suggestion. Looks like the alignment is the same as my commit.
Can you elaborate your change?

File: src/posix/Makefile-posix [hunk 4]
```diff
@@ -94,7 +94,10 @@ configure_OPTIONS                   = \
 # Platform specific switches
 
 ifeq ($(DAEMON),1)
-configure_OPTIONS              += --enable-posix-daemon
+configure_OPTIONS              += \
+    --enable-posix-daemon         \
+    LIBS=-lstdc++                 \
```
Comment (bukepo) [inline src/posix/Makefile-posix:99]: It seems this adds dependency on gcc. Will this change cause issues on platforms without gcc toolchain?

File: src/posix/platform/mainloop.hpp [hunk 5]
```diff
@@ -65,6 +65,12 @@ class Source
      */
     virtual void Process(const otSysMainloopContext &aContext) = 0;
 
+    /**
+     * This method marks desturctor virtual.
+     *
+     */
+    virtual ~Source() = default;
```
Comment (bukepo) [inline src/posix/platform/mainloop.hpp:72]: This is only necessary if the class is going to be deleted. see: https://patchwork.ozlabs.org/project/gcc/patch/BANLkTinxuAywtkOsNzSb+Ge6hmPWSm_D4Q@mail.gmail.com/

File: src/posix/Makefile-posix (hunk 4 above)
Comment (rxwen) [reply src/posix/Makefile-posix:99]: The stdc++ library is available for gcc and clang, on both ubuntu and macos.

File: src/posix/platform/mainloop.hpp (hunk 5 above)
Comment (rxwen) [reply src/posix/platform/mainloop.hpp:72]: Exactly.
My idea is it's a point of view on class design, it needs to be easier to be used right. If its destructor is virtual, the user of this polynomial class can worry less if it's used correctly.

File: src/posix/Makefile-posix (hunk 4 above)
Comment (bukepo) [reply src/posix/Makefile-posix:99]: it seems stdc++ is not availabe on Android platform. see https://developer.android.com/ndk/guides/cpp-support

**Review by superwhd — APPROVED** [review]
LGTM 👍. Thanks!

**Comment by rxwen** [comment]
@superwhd @simonlingoogle Sorry to bother. I merged the master branch in to my PR. It looks like the workflows requires approval to run and to complete the CI process. Would you take a look see if you may approve?

**Comment by jwhui** [comment]
> @superwhd @simonlingoogle Sorry to bother. I merged the master branch in to my PR. It looks like the workflows requires approval to run and to complete the CI process. Would you take a look see if you may approve?

@rxwen , I just approved the CI checks. Thanks.

**Comment by rxwen** [comment]
Thanks.
The ci failed in this step (https://github.com/openthread/openthread/runs/6988618609?check_suite_focus=true). The error seems not related to the PR I submitted. It's an error in the python code of commission test. Can you shed some lights on how to deal with this failure?

**Comment by jwhui** [comment]
@rxwen , any plan to move this PR forward?

**Comment by rxwen** [comment]
> @rxwen , any plan to move this PR forward?

Hi Jonathan,

I've merged the latest upstream code and tried again. Will checking the CI status closely.

**Comment by rxwen** [comment]
@jwhui Hi Jonathan, all the CI has passed. May I know how would it be get approved?

File: src/posix/Makefile-posix (hunk 4 above)
Comment (jwhui) [reply src/posix/Makefile-posix:99]: @rxwen , any thoughts on this?

Comment (rxwen) [reply src/posix/Makefile-posix:99]: I'm not quite sure about compiling with android ndk. Please wait me try it first.

File: src/posix/platform/mainloop.hpp [hunk 6]
```diff
@@ -65,6 +65,12 @@ class Source
      */
     virtual void Process(const otSysMainloopContext &aContext) = 0;
 
+    /**
+     * This method marks desturctor virtual method.
+     *
+     */
+    virtual ~Source() = default;
```
Comment (simonlingoogle) [inline src/posix/platform/mainloop.hpp:72]: Can we define the destructor to be private so that subclasses can't override it?

Comment (rxwen) [reply src/posix/platform/mainloop.hpp:72]: My thought on a destructor of a base class is it should be inheritable so that derived class can take advantage of it to do its own cleanup logic. If it's made private, I guess the derived will not be able to do so?

Comment (bukepo) [reply src/posix/platform/mainloop.hpp:72]: Actually, this isn't a problem. And the compiler will raise error if there is a real misuse, i.e. delete subclass with base class pointer.

Comment (rxwen) [reply src/posix/platform/mainloop.hpp:72]: > 

Agree on your point of having to link with stdc++. I will dig deeper on the dependency graph again why the lib is required.

Comment (rxwen) [reply src/posix/platform/mainloop.hpp:72]: > 

But i thinks polymorphism is designed to allow operating on subclass with base class pointer.

File: src/posix/Makefile-posix [hunk 7]
```diff
@@ -94,7 +94,9 @@ configure_OPTIONS                   = \
 # Platform specific switches
 
 ifeq ($(DAEMON),1)
-configure_OPTIONS              += --enable-posix-daemon
+configure_OPTIONS              += \
+    --enable-posix-daemon         \
+    $(NULL)
```
Comment (simonlingoogle) [inline src/posix/Makefile-posix:99]: unrelated change?

**Review by simonlingoogle — APPROVED** [review]
LGTM 👍 
with a nit

File: src/posix/Makefile-posix (hunk 7 above)
Comment (rxwen) [reply src/posix/Makefile-posix:99]: Good catch. It was introduced in previous commit for bringing in new link option. It should be reverted back to one line format.

**Review by bukepo — APPROVED** [review]

**Review by jwhui — APPROVED** [review]

**Comment by Ashish-Vara** [comment]
@rxwen, 

On this commit ID, I am getting same error which you have mentioned and added LIBS=-lstdc++. 

I see later you removed this again as no longer needed. May I know what is the reason for it? I am unable to compile without lstdc++ linking on this commit.

**Comment by Ashish-Vara** [comment]
HI @rxwen, 

We are building for i.MX platform using cross-tool chain from U20 machine. 

Error logs are:
/opt/fsl-imx-xwayland/5.15-kirkstone/sysroots/x86_64-pokysdk-linux/usr/libexec/aarch64-poky-linux/gcc/aarch64-poky-linux/11.2.0/real-ld: ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-backbone.o): in function `ot::Posix::MulticastRoutingManager::~MulticastRoutingManager()':
/home/development/openthread/ashish/bdb93829763_ot/openthread/src/posix/../../src/posix/platform/multicast_routing.hpp:51: undefined reference to `operator delete(void*)'
/opt/fsl-imx-xwayland/5.15-kirkstone/sysroots/x86_64-pokysdk-linux/usr/libexec/aarch64-poky-linux/gcc/aarch64-poky-linux/11.2.0/real-ld: ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-udp.o): in function `ot::Posix::Udp::Get()':
/home/development/openthread/ashish/bdb93829763_ot/openthread/src/posix/../../src/posix/platform/udp.cpp:611: undefined reference to `__cxa_guard_acquire'
/opt/fsl-imx-xwayland/5.15-kirkstone/sysroots/x86_64-pokysdk-linux/usr/libexec/aarch64-poky-linux/gcc/aarch64-poky-linux/11.2.0/real-ld: /home/development/openthread/ashish/bdb93829763_ot/openthread/src/posix/../../src/posix/platform/udp.cpp:611: undefined reference to `__cxa_guard_release'
/opt/fsl-imx-xwayland/5.15-kirkstone/sysroots/x86_64-pokysdk-linux/usr/libexec/aarch64-poky-linux/gcc/aarch64-poky-linux/11.2.0/real-ld: ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-udp.o): in function `ot::Posix::Udp::~Udp()':
/home/development/openthread/ashish/bdb93829763_ot/openthread/src/posix/../../src/posix/platform/udp.hpp:37: undefined reference to `operator delete(void*)'
/opt/fsl-imx-xwayland/5.15-kirkstone/sysroots/x86_64-pokysdk-linux/usr/libexec/aarch64-poky-linux/gcc/aarch64-poky-linux/11.2.0/real-ld: ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-daemon.o): in function `ot::Posix::Daemon::Get()':
/home/development/openthread/ashish/bdb93829763_ot/openthread/src/posix/../../src/posix/platform/daemon.cpp:364: undefined reference to `__cxa_guard_acquire'
/opt/fsl-imx-xwayland/5.15-kirkstone/sysroots/x86_64-pokysdk-linux/usr/libexec/aarch64-poky-linux/gcc/aarch64-poky-linux/11.2.0/real-ld: /home/development/openthread/ashish/bdb93829763_ot/openthread/src/posix/../../src/posix/platform/daemon.cpp:364: undefined reference to `__cxa_guard_release'
/opt/fsl-imx-xwayland/5.15-kirkstone/sysroots/x86_64-pokysdk-linux/usr/libexec/aarch64-poky-linux/gcc/aarch64-poky-linux/11.2.0/real-ld: ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-daemon.o): in function `ot::Posix::Daemon::~Daemon()':
/home/development/openthread/ashish/bdb93829763_ot/openthread/src/posix/../../src/posix/platform/daemon.hpp:39: undefined reference to `operator delete(void*)'
collect2: error: ld returned 1 exit status
Makefile:678: recipe for target 'ot-daemon' failed
make[5]: *** [ot-daemon] Error 1
Makefile:830: recipe for target 'all-recursive' failed
make[4]: *** [all-recursive] Error 1
Makefile:441: recipe for target 'all-recursive' failed
make[3]: *** [all-recursive] Error 1
Makefile:595: recipe for target 'all-recursive' failed
make[2]: *** [all-recursive] Error 1
Makefile:521: recipe for target 'all' failed
make[1]: *** [all] Error 2

**Comment by rxwen** [comment]
Is it a cross compilation toolchain based on GCC?
Can you also show the commit id of openthread that you'd building?
